src/dftbp/dftbplus/mainapi.F90

```diff
@@ -92,6 +95,167 @@ subroutine setGeometry(env, main, coords, latVecs, coordOrigin)
+    real(dp), intent(in) :: coordNeighbours(:,:)
+
+    !> mapping between neighbour reference and atom index in the central cell
+    integer, intent(in) :: neighbour2CentCell(:)
+
+    real(dp), allocatable :: dist2(:)
+    real(dp), pointer :: rCellVec(:)
+    integer :: nMaxNeighbours, nCellVec
+    integer :: iAtom, iCell, iNeigh, iImage, iCellVec, iAtFirst, iAtLast
+    integer, allocatable :: indx(:)
+    logical :: copyData
+
+    if (allocated(main%electronDynamics)) then
+      call error("Not implemented: Cannot set the neighbour list when time propagation is enabled")
+    end if
+    if (main%tLocalCurrents) then
+      call error("Not implemented: Cannot set the neighbour list when local bond-currents should be&
+          & computed")
+    end if
+
+    nMaxNeighbours = maxval(nNeighbour)
+
+    main%neighbourList%setExternally = .true.
+
+    #:if WITH_MPI
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .true., env)
+    #:else
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .false.)
+    #:endif
+
+    @:ASSERT(size(nNeighbour) == main%nAtom)
+    main%neighbourList%nNeighbour(:) = nNeighbour(:)
+    main%neighbourList%cutoff = cutOff
+
+    main%nAllAtom = size(neighbour2CentCell) + main%nAtom
+    @:ASSERT(size(neighbour2CentCell) == size(coordNeighbours, dim=2))
+
+    if (size(main%img2CentCell) /= main%nAllAtom) then
+      call reallocateArrays1(main%img2CentCell, main%iCellVec, main%coord, main%nAllAtom)
+    end if
+
+    !> Prepend data for the atoms in the central cell
+    main%coord(1:3, 1:main%nAtom) = main%coord0(1:3, 1:main%nAtom)
+    main%img2CentCell(1:main%nAtom) = [(iAtom, iAtom = 1, main%nAtom)]
+    main%iCellVec(1:main%nAtom) = 1
+
+    if (main%nAtom < main%nAllAtom) then
+      main%coord(1:3,main%nAtom+1:) = coordNeighbours(1:3,:)
+      main%img2CentCell(main%nAtom+1:) = neighbour2CentCell
+      main%iCellVec(main%nAtom+1:) = 0
+
+      !> Now set main%iCellVec: Iterate over all cells, calculate the coordinates the atom would
+      !> have there, and determine the cell in which this is very close to the actual coordinates
+      nCellVec = size(main%rCellVec, dim=2)
+      allocate(dist2(nCellVec))
+
+      rCellVec(1:3*nCellVec) => main%rCellVec
+
+      call distributeRangeInChunks(env, main%nAtom + 1, main%nAllAtom, iAtFirst, iAtLast)
+
+      do iAtom = iAtFirst, iAtLast
+        iImage = main%img2CentCell(iAtom)
+        do iCell = 1, nCellVec
+          iCellVec = 3 * (iCell - 1)
+          dist2(iCell) = (main%coord0(1,iImage) + rCellVec(iCellVec + 1) -&
+              & main%coord(1,iAtom))**2 +&
+              & (main%coord0(2,iImage) + rCellVec(iCellVec + 2) - main%coord(2,iAtom))**2 +&
+              & (main%coord0(3,iImage) + rCellVec(iCellVec + 3) - main%coord(3,iAtom))**2
+        end do
+        main%iCellVec(iAtom) = minloc(dist2, dim=1)
+      end do
+
+      call assembleChunks(env, main%iCellVec)
```

Comment:
```suggestion
      call assembleChunks(env, main%iCellVec(main%nAtom + 1 :))
```
I think, you get wrong `iCellVec` indices for the central cell atoms otherwise.

Reply:
Yes, you are right. I don't know why my tests did not show an incorrect behaviour caused by this bug.